symengine/polys.cpp

```diff
@@ -245,8 +246,54 @@ RCP<const Polynomial> neg_poly(const Polynomial &a) {
 }
 
 #ifdef HAVE_SYMENGINE_PIRANHA
-RCP<const Polynomial> mul_poly(RCP <const Polynomial> a, RCP <const Polynomial> b) {
-    hash_set C, A = a->polys_set_, B = b->polys_set_;
+void _normalise_polymul(RCP <const Polynomial> &p, RCP <const Polynomial> &q, vec_symbol vars_,
+    Polynomial &p_new, Polynomial &q_new) {
+    using ka = piranha::kronecker_array<long long>;
+    vars_ = p->vars_;
+    int c_1 = p->vars_.size();
+    int c_2 = 0;
+    for (auto i = q->vars_.begin(); i < q->vars_.end(); i++) {
+        for (auto j = p->vars_.begin(); i < p->vars_.end(); i++) {
+            if ((*i)->get_name() != (*j)->get_name()) {
+                vars_.push_back(*i);
+                c_2++;
+            }
+        }
+    }
+    for (auto &a: p->polys_set_) {
+        std::vector<long long> old(p->vars_.size());
+        m_pair temp;
+
+        ka::decode(old, a.first);
+        for (int i = 0; i < c_2; i++) {
+            old.push_back(0);
+        }
+
+        temp.first = ka::encode(old);
+        temp.second = a.second;
+        p_new.polys_set_.insert(temp);
+    }
+    for (auto &a: q->polys_set_) {
+        if (c_1 == 0) {
+            q_new.polys_set_ = q->polys_set_;
+        } else {
+            std::vector<long long> new_(c_1);
+            std::vector<long long> old(q->vars_.size());
+            m_pair temp;
+
+            ka::decode(old, a.first);
+            for (unsigned int i = 0; i < old.size(); i++) {
+                new_.push_back(old[i]);
```

Comment:
Is `q->vars_.size()` and `old.size()` the same? Otherwise you are probably accessing out of bounds. We need to enable STL Debug checks (https://github.com/sympy/symengine/issues/307).